File: Reqnroll/Chapter16/16.1-ScenarioResourcePattern/After/WIMP.Specs/Support/Hooks.cs
using Reqnroll;

using WIMP.Specs.Drivers;

namespace WIMP.Specs.Support;

[Binding]
public class Hooks(AppHostingContext appHostingContext)
{
    [BeforeScenario("@payment_gateway", Order = 1)]
    public void InitializePaymentGateway()
    {
        appHostingContext.PaymentGateway = PaymentGatewaySimulator.Start();
    }

    [BeforeScenario(Order = 2)]
    public void CreateAppHost()
    {
        appHostingContext.AppHost = new WimpAppHost(appHostingContext.PaymentGateway);
    }

    [AfterScenario]
    public void DisposeAppHost()
    {
        appHostingContext.AppHost.Dispose();
    }

    [AfterScenario("@payment_gateway")]
    public void DisposePaymentGateway()
    {
        appHostingContext.PaymentGateway?.Stop();
    }
}

File: CucumberJVM/Chapter20/20.4-ParameterizedScenarioPattern/After/src/test/java/com/wimp/app/specs/support/Hooks.java
package com.wimp.app.specs.support;

import com.wimp.app.specs.drivers.DatabaseDriver;
import com.wimp.app.specs.drivers.PaymentGatewayScenarioContextProxy;
import com.wimp.app.specs.drivers.PaymentGatewaySimulator;
import io.cucumber.java.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Hooks {
    private static final Logger log = LoggerFactory.getLogger(Hooks.class);

    private final PaymentGatewaySimulator paymentGatewaySimulator;
    private final PaymentGatewayScenarioContextProxy paymentGatewayScenarioContextProxy;
    private final DatabaseDriver databaseDriver;
    private final TestConfigurationProvider testConfigurationProvider;

    public Hooks(PaymentGatewaySimulator paymentGatewaySimulator, PaymentGatewayScenarioContextProxy paymentGatewayScenarioContextProxy, DatabaseDriver databaseDriver, TestConfigurationProvider testConfigurationProvider) {
        this.paymentGatewaySimulator = paymentGatewaySimulator;
        this.paymentGatewayScenarioContextProxy = paymentGatewayScenarioContextProxy;
        this.databaseDriver = databaseDriver;
        this.testConfigurationProvider = testConfigurationProvider;
    }

    @Before(order = 0)
    public void resetDatabase() {
        log.info("Test database: use-stub={}", testConfigurationProvider.testDatabase().useStub());
        databaseDriver.emptyDatabase();
    }

    @Before(value = "@payment_gateway", order = 1)
    public void initializePaymentGateway() {
        var paymentGateway = paymentGatewaySimulator.start();
        paymentGatewayScenarioContextProxy.setPaymentGatewayOverride(paymentGateway);
    }
}
